FrequencyConversion/FC_solve_EoM.py

```diff
@@ -0,0 +1,237 @@
+        L (float): length of propagation in the crystal
+        D (float): number depicting the strength of the nonlinear interaction
+
+    Returns:
+        (array, :logm of the propagator
+        array)  :propagator
+    """
+
+    # Upper left and bottom right (diagonal) blocks
+    ka = 1 / va - 1 / vp
+    kc = 1 / vc - 1 / vp
+
+    Ka = np.diag(1j * dw * ka)
+    Kc = np.diag(1j * dw * kc)
+    # Bottom left and upper right (non-diagonal) blocks
+
+    J = np.block([[Ka, -1j * np.conj(D) * pump], [-1j * D * pump.conj().T, Kc]])
+    prop = expm(J * L)
+    return J, prop
+
+
+def check_prop(prop, print_checks=False):
+    """
+    Verifies that the obtained propagator fulfills the 6 conditions imposed in the paper
+    by Andreas Christ from 2013 (equations A3-4 and A6-7).
+    Returns the maximum error for each condition.
+
+    Args :
+        prop (matrix): the propagator to perform the checks on
+        print_checks (bool): if True, prints if the condition is True or False. False by default.
+
+    Returns :
+        array: maximum value of the absolute value of the condition mismatch,
+               for each condition (always an array of 6 floats)
+    """
+    Ua, Va, Vc, Uc = get_blocks(prop)
+    Vc = -Vc
+    N = Ua.shape[0]
+    cond1a = Ua @ dag(Ua) + Va @ dag(Va)
+    cond1b = Uc @ dag(Uc) + Vc @ dag(Vc)
+    cond3a = dag(Ua) @ Ua + dag(Vc) @ Vc
+    cond3b = dag(Uc) @ Uc + dag(Va) @ Va
+    cond2 = Ua @ dag(Vc) - Va @ dag(Uc)
+    cond4 = dag(Ua) @ Va - dag(Vc) @ Uc
+    I = np.identity(N)
+    O = np.zeros((N, N))
+    conds = np.array([cond1a, cond1b, cond2, cond3a, cond3b, cond4])
+    comps = np.array([I, I, O, I, I, O])
+    labels = ["1a", "1b", "2", "3a", "3b", "4"]
+    to_return = np.array([])
+    for i in range(len(labels)):
+        cond = conds[i]
+        comp = comps[i]
+        txt = "Condition {name} is {TF}"
+        if print_checks:
+            print(txt.format(name=labels[i], TF=np.allclose(cond, comp)))
+        to_return = np.append(to_return, np.max(np.abs(cond - comp)))
+    return to_return
+
+
+def remove_phases(prop, dw, L, vp, va, vc):
+    """
+    Removes the free space propagation phases from the propagator.
+
+    Args:
+        prop (array): propagator with phases
+        dw (array): frequency vector
+        L (float): length of propagation in crystal
+        vp (float): group velocity for pump
+        va (float): group velocity for a beam
+        vc (float): group velocity for c beam
+
+    Returns:
+        array :the phase-free propagator as a matrix
+    """
+    kc = 1 / vp - 1 / vc
+    ka = 1 / va - 1 / vp
+
+    # separate into blocks
+    Ua, Va, Vc, Uc = get_blocks(prop)
+    # remove phases
+    Uc = np.diag(np.exp(1j * dw * kc * L)) @ Uc
+    Vc = np.diag(np.exp(1j * dw * kc * L)) @ Vc
+    Va = np.diag(np.exp(-1j * dw * ka * L)) @ Va
+    Ua = np.diag(np.exp(-1j * dw * ka * L)) @ Ua
+    return np.block([[Ua, Va], [Vc, Uc]])
+
```

Comment:
Thsi removal of phases in incorrect. You need to multiply phases on both sides.

Reply:
As Nico said, this is incorrect. The correct method is outlined in  https://journals.aps.org/pra/abstract/10.1103/PhysRevA.102.033519, paragraph between Eq.65 and 66. I also do it in my "phases" function in 'propagator.py'. Note that these are for SPDC where the idler mode is ^\dagger. For F.C. one needs to take the opposite sign for that mode.

Reply:
Yes, I remember thinking about that and looking at it when looking for how to remove the phases. But then, I thought that since I have my waveguide start at z=0 and finish at z=L I only had phases on one side, since my z_0 would be 0. Is that incorrect or is there a profound reason that I should have z_0 = -L/2, z_1 = L/2, or is it just a convention? In any case, I can change it for sure, just so it remains consistent with what you have in propagator.py.

Reply:
Right ok, yeah we should keep it all consistent.